infra/is-ngcc-necessary.ts

```diff
@@ -0,0 +1,58 @@
+import * as chalk from 'chalk';
+import { existsSync, readFileSync } from 'fs';
+import { resolve } from 'path';
+
+const packages = Object.keys(JSON.parse(readFileSync(resolve(__dirname, '../package.json'), 'utf8')).dependencies);
+const knownNonNgccPackages: string[] = JSON.parse(readFileSync(resolve(__dirname, './non-ngcc-packages.json'), 'utf8'));
+const knownNonNgccPackageMatchers = knownNonNgccPackages.map((pattern: string) => new RegExp(`^${pattern}$`));
+
+console.log(chalk.gray('Checking for packages that no longer need ngcc processing'));
+
+const nonNgccPackages = [];
+for(const packageName of packages) {
+  if (knownNonNgccPackageMatchers.some(matcher => matcher.test(packageName))) {
+    continue;
+  }
+  const packagePath = resolve(__dirname, '../node_modules', packageName);
+  const packageJson = JSON.parse(readFileSync(resolve(packagePath, 'package.json'), 'utf8'));
+  const typingsPath = packageJson.typings || packageJson.types || guessTypingsFromPackageJson(packagePath, packageJson);
+
+  if (!typingsPath) {
```

Comment:
Is it guaranteed that packages without typings will not include metadata JSON files (or is it just a heuristic)?

Reply:
Technically `types` and `typeVersions` are valid as well - does ngcc just ignore those?

Reply:
@alxhub - this is the logic used by ngcc (copied here):

```ts
const typingsPath = packageJson.typings || packageJson.types || guessTypingsFromPackageJson(packagePath, packageJson);
```

So we try `typings` first, then `types`, then fallback on looking for a `,d,ts` file next to one of the JS files.

Reply:
I had not even heard of the `typeVersions` property before today!
Generally ViewEngine libraries were built with ng-packagr, so they did not use this property.

So I think we are fine here. In any case, we are not likely to add "new" packages to this project. Only remove the ones that we find no longer need ngcc. In the worst case, this script will fail to report a package that updates to no longer need ngcc because it has started using `typeVersions` rather than `types` or `typings`.